proposals/TP3-xds-error-propagation.md

```diff
@@ -0,0 +1,122 @@
+
+This proposal includes a new field for each subscribed Resource, called `ResourceError` which will provide detailed information for resource specific issues. The client must use this additional field to obtain notification for resources the xDS Management server couldn’t procure and provide necessary notification to the application. 
+
+## Background
+
+A frequent use case for xDS involves client subscription to multiple resources from the management server. Currently, if the xDS management server cannot provide a subset of these resources, the client experiences a complete loss of visibility. This can occur for various reasons, including but not limited to potential permission issues. This lack of granularity in the response can pose significant operational challenges for general client observability or debugging. 
+
+The xDS protocol does have a way for xDS clients to NACK responses back from the xDS Management server. The NACK response also contains an `error_detail` field which the Management Server can use to extract further information about the rejection. But this NACK’ing mechanism is restricted to the client i.e. there is no way for Management Server to actually convey a notification(Ex: unavailability, permission issues etc) regarding some or all the resource requests that are being subscribed by the client. This eventually leads to the client timeouts which, although it conveys an error back to the application, it misses the actual context for the issue. In most cases the xDS Management Servers might not even be accessible for the client applications to debug these issues without escalation. 
+
+### Related Proposals:
+
+N/A
+
+## Proposal
+
+The proposal introduces two new fields to the DiscoveryResponse proto. The first, `resource_errors`, will detail Resource-specific errors and be included in both SotW and Incremental xDS protocols. The second, `removed_resources`, will specifically identify resources that no longer exist, addressing a current gap in the SotW protocol's signaling capabilities. 
+
+```textproto
+// New Proto Message
+// Contains detailed error detail for a single resource
+message ResourceError {
+ ResourceName resource_name = 1;
+
+ google.rpc.Status error_detail = 2;
+}
+
+message DiscoveryResponse {
+
+// The version of the response data.
+string version_info = 1;
+
+….
+…
+
+
+ // The control plane instance that sent the response.
+  config.core.v3.ControlPlane control_plane = 6;
+
+// NEW_FIELD
+// An optional repeated list of error details that the control plane 
+// can provide to notify the client of issues for the resources that 
+// are subscribed.
+// This allows the xDS management server to provide optional 
+// notifications in case of unavailability, permissions errors 
+// without the client having to wait for the `config fetch` timeout.
+repeated ResourceError resource_errors = N;
+
+// NEW FIELD
+// For non-existing resources.
+repeated ResourceName removed_resources = N;
```

Comment:
Isn't this implied by the lack of existence in `resources`?

Reply:
Currently if the xDS management server doesn't provide any resources its not clear if the resources are not present or if there was just a delay in sending these resources across to the client. The client currently can't differentiate between the two scenarios. 

By explicitly providing a list of resources that don't exist, the client can stop the timer and conclude that the resources are missing.

Reply:
I mean I kind of get it.. but also this has nothing to do with error propagation really? It feels like it's slipping in "retrofit sotw to have some delta functionality" into an error propagation proposal when they are really orthogonal

Reply:
The major reason for this is to `propagate the error` that the resource is technically missing. If we don't cover this scenario we still have this gap for SotW. 

I do resonate with your point that this might be overstepping the scope of this error propagation proposal. I am ok adding this into another specific xRFC.

@markdroth in case he has additional thoughts.

Reply:
I can see how this might seem tangential, but I think it's an important part of this proposal.  The overall goal here is for the control plane to be able to explicitly indicate errors to the client, so that the client can tell the difference between various types of errors -- e.g., I think we want to understand the difference between "does not exist" and "permission denied" and (for SotW) "the control plane is just being slow".

That having been said, in our internal discussions, we did go back and forth a few times as to whether we needed a separate field for does-not-exist or whether we should just encode an error with status NOT_FOUND in that case.  I think the core question here is whether we'd want the client to react in a different way to does-not-exist than to an error like permission-denied.  Today, gRPC does have special handling for does-not-exist: it causes us to deconfigure the channel instead of just sticking with the previously cached resource, if any, which we would on other types of errors (e.g., a connectivity error with the control plane or receiving an invalid resource).  But it's worth noting that today, the control plane would need to treat permission-denied as does-not-exist, so the client would use that same special behavior in this case.  I guess we should explicitly decide what behavior we want for cases like permission-denied.

So I see a few options here:

1. gRPC treats any error from the control plane the same as does-not-exist, just with a better error message.  In this case, we don't really need the separate `removed_resources` field, because we can encode both cases in the `resource_errors` field.
2. gRPC treats does-not-exist differently than other errors.  We have two ways of reflecting this in the transport protocol:
   a. We can encode does-not-exist via this separate `removed_resources` field, as this xRFC currently indicates.
   b. We can not add the `removed_resources` field and instead rely on NOT_FOUND status in `resource_errors`.  This seems a little sub-optimal, because it imbues one particular status code with special meaning, which is a lot less clear.

@ejona86 and @dfawley, thoughts...?

Reply:
Done, changed